tests/trainer/logging/test_eval_loop_logging_1_0.py

```diff
@@ -321,3 +322,88 @@ def test_monitor_val_epoch_end(tmpdir):
+
+    loss_values, patience, expected_stop_epoch = ([6, 5, 5, 5, 5, 5], 3, 4)
+
+    class TestCallback(callbacks.Callback):
+
+        callback_funcs_called = []
+
+        def on_validation_start(self, trainer, pl_module):
+            func_name = 'on_validation_start'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+        def on_validation_epoch_start(self, trainer, pl_module):
+            func_name = 'on_validation_epoch_start'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+        def on_validation_batch_start(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+            func_name = 'on_validation_batch_start'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+        def on_validation_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
+            func_name = 'on_validation_batch_end'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+        def on_validation_epoch_end(self, trainer, pl_module):
+            func_name = 'on_validation_epoch_end'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+        def on_validation_end(self, trainer, pl_module):
+            func_name = 'on_validation_end'
+            pl_module.log(func_name, 0)
+            self.callback_funcs_called.append(func_name)
+
+
+
+    class TestModel(BoringModel):
+
+        validation_return_values = torch.Tensor(loss_values)
+        count = 0
+
+        def validation_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.log('c', self.count)
+            return {"x": loss}
+
+        def validation_epoch_end(self, outputs):
+
+            loss = self.validation_return_values[self.count]
+            self.count += 1
+            self.log("val_loss", loss)
+            return {"val_loss": loss}
+
+    max_epochs = 5
+    model = TestModel()
+    # using Early Stopping to make sure callback_metrics is correctly set for `on_validation_epoch_end` hook
+    early_stop_callback = pl.callbacks.EarlyStopping(monitor="val_loss", patience=3, verbose=True)
+    test_callback = TestCallback()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=0,
+        limit_val_batches=4,
+        limit_test_batches=0,
+        val_check_interval=1.0,
+        num_sanity_val_steps=0,
+        max_epochs=max_epochs,
+        callbacks=[early_stop_callback, test_callback]
+    )
+    trainer.fit(model)
+
+    expected_logged_metrics = set(test_callback.callback_funcs_called + ["c", "val_loss"])
+    logged_metrics = set(trainer.logged_metrics.keys())
+    assert logged_metrics == expected_logged_metrics, logged_metrics
+    assert trainer.current_epoch == expected_stop_epoch
```

Comment:
can you also assert the number of times the metrics were called per hook name?

Reply:
this is also conflating two tests: early stopping and logging.
Can you separate both tests to keep things simple?

One test for the new changes to the metrics.
One test for early stopping (is this even necessary? don't we have a ton of these tests already?)

Reply:
Yes, I let me remove it. It was to make sure the callback_metrics was properly set. Mainly for debugging purpose.
Also, there is an inconsistency with on_batch_start, on_batch_end